app/controllers/omniauth_callbacks_controller.rb

```diff
@@ -0,0 +1,21 @@
+class OmniauthCallbacksController < Devise::OmniauthCallbacksController
+  skip_authorization_check
+
+  def headquarters
+    user = User.from_omniauth(auth_hash)
+
+    if user.persisted?
+      sign_in_and_redirect user, event: :authentication
+      set_flash_message(:notice, :success, kind: :headquarters) if is_navigational_format?
+    else
+      session['devise.headquarters_data'] = request.env['omniauth.auth']
```

Comment:
No, it's from devise. It was copied from their example in the docs. https://github.com/plataformatec/devise/wiki/OmniAuth:-Overview